subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/rocache/ReadOnlyDependencyCacheWithinContainerTest.groovy

```diff
@@ -0,0 +1,170 @@
+        // when we build it for R/O cache and when we use it in the build in a
+        // container
+        expectRefresh(core)
+        expectRefresh(utils)
+
+        def result = container
+            .withBuildDir(testBuildDir)
+            .succeeds("resolve")
+
+        then:
+        result.with {
+            assertOutputContains("/gradle-home/caches/modules-2/files-2.1/org.readonly/core/1.0/")
+            assertOutputContains("/gradle-home/caches/modules-2/files-2.1/org.readonly/util/1.0/")
+        }
+    }
+
+    def "can use a read-only cache within multiple containers concurrently"() {
+        given:
+        def ids = (0..3)
+        exposeServerToContainers()
+        def containers = ids.collect { id ->
+            // The HEAD requests are because the URL of the repository is different
+            // when we build it for R/O cache and when we use it in the build in a
+            // container
+            expectRefresh(core)
+            expectRefresh(utils)
+            def testBuildDir = createContainerBuild("build$id")
+            newContainer()
+                .withBuildDir(testBuildDir)
+        }
+
+        synchronizer.expectConcurrent(ids.collect { "build$it".toString() })
+
+        when:
+        def results = Collections.synchronizedCollection([])
+        containers.collect { container ->
+            Thread.start {
+                results << container.succeeds("resolve")
+            }
+        }*.join()
+
+        then:
+        results.each { result ->
+            result.with {
+                assertOutputContains("/gradle-home/caches/modules-2/files-2.1/org.readonly/core/1.0/")
+                assertOutputContains("/gradle-home/caches/modules-2/files-2.1/org.readonly/util/1.0/")
+            }
+        }
+
+    }
+
+    private static void expectRefresh(MavenHttpModule module) {
+        module.pom.expectHead()
+        module.pom.sha1.expectGet()
+        module.moduleMetadata.expectHead()
+        module.moduleMetadata.sha1.expectGet()
+        module.artifact.expectHead()
+        module.artifact.sha1.expectGet()
+    }
+
+    private void exposeServerToContainers() {
+        GradleInContainer.exposeHostPort(server.port)
+    }
+
+    private TestFile createContainerBuild(String id = 'build') {
+        def testBuildDir = temporaryFolder.createDir("test-build-$id")
+        testBuildDir.file("settings.gradle") << """
+            rootProject.name = "test-build-in-container"
+        """
+        testBuildDir.file("build.gradle") << """
+            plugins {
+                id 'java-library'
+            }
+
+            repositories {
+                maven {
+                   allowInsecureProtocol = true
+                   url "http://host.testcontainers.internal:${server.port}/repo"
+                }
+            }
+
+            dependencies {
+                implementation "org.readonly:core:1.0"
+            }
+
+            tasks.register("resolve") {
```

Comment:
not related to the change, but:

It would be cool to have an API in Gradle to resolve **all** dependencies - very helpful when building Docker images

Reply:
In short: you can't. Gradle cannot know in advance all configurations created by all tasks. It cannot know about detached configurations or configurations resolved at task execution time. For dependency verification we have an _approximation_ of this, but it's not sufficient so you'll always have missing dependencies.